File: Leetcode/Medium/63. Unique Paths II.cpp
class Solution
{
public:
    int uniquePathsWithObstacles(vector<vector<int>> &obstacleGrid)
    {
        int m = obstacleGrid.size(), n = obstacleGrid[0].size();
        if (obstacleGrid[0][0] == 1 || obstacleGrid[m - 1][n - 1] == 1)
            return 0;
        obstacleGrid[0][0] = 1;
        for (int r = 1; r < m; r++)
        {
            obstacleGrid[r][0] = (obstacleGrid[r][0] == 0 && obstacleGrid[r - 1][0] == 1) ? 1 : 0;
        }
        for (int c = 1; c < n; c++)
        {
            obstacleGrid[0][c] = (obstacleGrid[0][c] == 0 && obstacleGrid[0][c - 1] == 1) ? 1 : 0;
        }
        for (int r = 1; r < m; r++)
        {
            for (int c = 1; c < n; c++)
            {
                if (obstacleGrid[r][c] == 1)
                {
                    obstacleGrid[r][c] = 0;
                }
                else
                {
                    obstacleGrid[r][c] = obstacleGrid[r - 1][c] + obstacleGrid[r][c - 1];
                }
            }
        }
        return obstacleGrid[m - 1][n - 1];
    }
};

File: Leetcode/Medium/63. Unique Paths II/63. Unique Paths II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        if obstacleGrid[0][0] == 1 or obstacleGrid[m - 1][n - 1] == 1:
            return 0
        
        obstacleGrid[0][0] = 1

        for r in range(1, m):
            obstacleGrid[r][0] = 0 if ((obstacleGrid[r][0] == 1) or (obstacleGrid[r - 1][0] == 0)) else 1

        for c in range(1, n):
            obstacleGrid[0][c] = 0 if (obstacleGrid[0][c] == 1 or obstacleGrid[0][c - 1] == 0) else 1

        for r in range(1, m):
            for c in range(1, n):
                if obstacleGrid[r][c] == 1:
                    obstacleGrid[r][c] = 0
                else:
                    obstacleGrid[r][c] = obstacleGrid[r - 1][c] + obstacleGrid[r][c - 1]

        print (obstacleGrid)
        return obstacleGrid[m - 1][n - 1]
